internal/managementrouter/user_defined_alert_rule_create.go

```
@@ -0,0 +1,75 @@
+package managementrouter
+
+import (
+	"encoding/json"
+	"net/http"
+	"strings"
+
+	monitoringv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/openshift/monitoring-plugin/pkg/management"
+)
+
+type AlertingRulePayload struct {
+	Alert       string            `json:"alert,omitempty"`
+	Expr        string            `json:"expr,omitempty"`
+	For         string            `json:"for,omitempty"`
+	Labels      map[string]string `json:"labels,omitempty"`
+	Annotations map[string]string `json:"annotations,omitempty"`
+}
+
+type PrometheusRulePayload struct {
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+	GroupName string `json:"groupName,omitempty"`
+}
+
+type CreateAlertRuleRequest struct {
+	AlertingRule   *AlertingRulePayload   `json:"alertingRule,omitempty"`
+	PrometheusRule *PrometheusRulePayload `json:"prometheusRule,omitempty"`
+}
+
+type CreateAlertRuleResponse struct {
+	Id string `json:"id"`
+}
+
+func (hr *httpRouter) CreateUserDefinedAlertRule(w http.ResponseWriter, req *http.Request) {
+	var payload CreateAlertRuleRequest
+	if err := json.NewDecoder(req.Body).Decode(&payload); err != nil {
+		writeError(w, http.StatusBadRequest, "invalid request body")
+		return
+	}
+
+	if payload.AlertingRule == nil {
+		writeError(w, http.StatusBadRequest, "alertingRule is required")
+		return
+	}
+
+	rule := monitoringv1.Rule{
+		Alert:       payload.AlertingRule.Alert,
+		Expr:        intstr.FromString(payload.AlertingRule.Expr),
+		For:         monitoringv1.DurationPointer(payload.AlertingRule.For),
+		Labels:      payload.AlertingRule.Labels,
+		Annotations: payload.AlertingRule.Annotations,
+	}
+
+	if payload.PrometheusRule == nil {
+		writeError(w, http.StatusBadRequest, "prometheusRule is required")
+		return
+	}
```

move before `rule := monitoringv1.Rule{`
fails faster